File: Dot examples/ASP Book Examples/Book Examples/Chapter21/CS/SqlVirtualPathProviderApp/App_Code/SqlVirtualPathProvider.cs
using System;
using System.Web;
using System.Web.Caching;
using System.Collections;
using System.Collections.Generic;
using System.Web.Hosting;

namespace AspNetUnleashed
{

    public class SqlVirtualPathProvider : VirtualPathProvider
    {
        /// <summary>
        /// Register VirtualPathProvider for the application
        /// </summary>
        public static void AppInitialize()
        {
            SqlVirtualPathProvider sqlProvider = new SqlVirtualPathProvider();
            HostingEnvironment.RegisterVirtualPathProvider(sqlProvider);
        }


        public SqlVirtualPathProvider() : base() { }

        /// <summary>
        /// Returns true when the file is a virtual file
        /// instead of a normal filesystem file
        /// </summary>
        private bool IsVirtualFile(string virtualPath)
        {
            String appVirtualPath = VirtualPathUtility.ToAppRelative(virtualPath);
            return !appVirtualPath.StartsWith("~/admin/", StringComparison.InvariantCultureIgnoreCase);
        }

        /// <summary>
        /// Returns true when a file exists
        /// </summary>
        public override bool FileExists(string virtualPath)
        {
            if (IsVirtualFile(virtualPath))
                return VirtualFiles.FileExists(virtualPath);
            else
                return Previous.FileExists(virtualPath);
        }

        /// <summary>
        /// Gets a SqlVirtualFile which corresponds
        /// to a file with a certain path
        /// </summary>
        public override VirtualFile GetFile(string virtualPath)
        {
            if (IsVirtualFile(virtualPath)) 
                return new SqlVirtualFile(virtualPath); 
            else
                return Previous.GetFile(virtualPath);
        }

        /// <summary>
        /// Returns true when a directory exists
        /// </summary>
        public override bool DirectoryExists(string virtualPath)
        {
            if (IsVirtualFile(virtualPath))
                return VirtualFiles.DirectoryExists(virtualPath);
            else
                return Previous.DirectoryExists(virtualPath);
        }

        /// <summary>
        /// Returns a SqlVirtualDirectory which corresponds
        /// to a virtual path
        /// </summary>
        public override VirtualDirectory GetDirectory(string virtualPath)
        {
            if (IsVirtualFile(virtualPath))
                return new SqlVirtualDirectory(virtualPath);
            else
                return Previous.GetDirectory(virtualPath);
        }

        /// <summary>
        /// Gets the SqlCacheDependency object for the VirtualFilesDB
        /// database
        /// </summary>
        public override CacheDependency GetCacheDependency(string virtualPath, IEnumerable virtualPathDependencies, DateTime utcStart)
        {
            if (IsVirtualFile(virtualPath))
                return new SqlCacheDependency("VirtualFiles", "VirtualFiles");
            else
                return Previous.GetCacheDependency(virtualPath, virtualPathDependencies, utcStart);
        }

    }
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter34/vb/App_Code/VirtualPathProvider/SqlVirtualPathProvider.vb
Imports System
Imports System.Web
Imports System.Web.Caching
Imports System.Collections
Imports System.Collections.Generic
Imports System.Web.Hosting

Namespace Superexpert

    Public Class SqlVirtualPathProvider
     Inherits VirtualPathProvider
        '/ <summary>
        '/ Register VirtualPathProvider for the application
        '/ </summary>
        Public Shared Sub AppInitialize()
            Dim sqlProvider As SqlVirtualPathProvider = New SqlVirtualPathProvider()
            HostingEnvironment.RegisterVirtualPathProvider(sqlProvider)
        End Sub


        Public Sub New()
         MyBase.New()
        End Sub

        '/ <summary>
        '/ Returns true when the file is a virtual file
        '/ instead of a normal filesystem file
        '/ </summary>
        Private Function IsVirtualFile(ByVal virtualPath As String) As Boolean
            Dim appVirtualPath As String = VirtualPathUtility.ToAppRelative(virtualPath)
            Return appVirtualPath.StartsWith("~/virtual/", StringComparison.InvariantCultureIgnoreCase)
        End Function

        '/ <summary>
        '/ Returns true when a file exists
        '/ </summary>
        Public Overrides Function FileExists(ByVal virtualPath As String) As Boolean
            If IsVirtualFile(virtualPath) Then
                Return VirtualFiles.FileExists(virtualPath)
            Else
                Return Previous.FileExists(virtualPath)
            End If
        End Function

        '/ <summary>
        '/ Gets a SqlVirtualFile which corresponds
        '/ to a file with a certain path
        '/ </summary>
        Public Overrides Function GetFile(ByVal virtualPath As String) As VirtualFile
            If IsVirtualFile(virtualPath) Then
                Return New SqlVirtualFile(virtualPath)
            Else
                Return Previous.GetFile(virtualPath)
            End If
        End Function

        '/ <summary>
        '/ Returns true when a directory exists
        '/ </summary>
        Public Overrides Function DirectoryExists(ByVal virtualPath As String) As Boolean
            If IsVirtualFile(virtualPath) Then
                Return VirtualFiles.DirectoryExists(virtualPath)
            Else
                Return Previous.DirectoryExists(virtualPath)
            End If
        End Function

        '/ <summary>
        '/ Returns a SqlVirtualDirectory which corresponds
        '/ to a virtual path
        '/ </summary>
        Public Overrides Function GetDirectory(ByVal virtualPath As String) As VirtualDirectory
            If IsVirtualFile(virtualPath) Then
                Return New SqlVirtualDirectory(virtualPath)
            Else
                Return Previous.GetDirectory(virtualPath)
            End If
        End Function

        '/ <summary>
        '/ Gets the SqlCacheDependency object for the VirtualFilesDB
        '/ database
        '/ </summary>
        Public Overrides Function GetCacheDependency(ByVal virtualPath As String, ByVal virtualPathDependencies As IEnumerable, ByVal utcStart As DateTime) As CacheDependency
            If IsVirtualFile(virtualPath) Then
                Return New SqlCacheDependency("VirtualFiles", "CodeSample")
            Else
                Return Previous.GetCacheDependency(virtualPath, virtualPathDependencies, utcStart)
            End If
        End Function

    End Class
End Namespace

